### PR TITLE
Rename sys file, implement RW Bytes ops

### PR DESCRIPTION
**Pull request by r-rojo**
* Rename sys.h and sys.c to sysobject.h and sysobject.c
* Implement fpgaWriteObjectBytes, and fpgaReadObjectBytes functions
* Also change fn signature of fpgaWriteObjectBytes to use len output
param
* Rename gtunit in CMake to gtunit-sysobject

### CODE REVIEW

File: libopae/src/sysobject.c [hunk 1]
```diff
@@ -98,20 +101,38 @@ fpga_result __FPGA_API__ fpgaReadObjectBytes(fpga_token token, const char *key,
...
+				 strerror(errno));
+		} else {
+			FPGA_ERR("Offset after lseek < offset argument");
+		}
+		return FPGA_EXCEPTION;
+	}
+	*len = read(fd, buffer, *len);
 
-	return res;
+	return FPGA_OK;
```
Comment (asgibbons) [inline libopae/src/sysobject.c]: Need to close the file.

File: libopae/src/sysobject.c [hunk 2]
```diff
@@ -144,17 +165,37 @@ fpga_result __FPGA_API__ fpgaWriteObject64(fpga_handle handle, const char *key,
...
+			FPGA_ERR("Error with lseek operation: %s",
+				 strerror(errno));
+		} else {
+			FPGA_ERR("Offset after lseek < offset argument");
+		}
+		return FPGA_EXCEPTION;
+	}
+	*len = write(fd, buffer, *len);
+
+	return FPGA_OK;
```
Comment (asgibbons) [inline libopae/src/sysobject.c]: Need to close the file.

File: tests/unit/gtsys.cpp [hunk 3]
```diff
@@ -38,7 +38,7 @@ extern "C" {
 #include "gtest/gtest.h"
 #include "common_test.h"
 
-TEST(sys, read32) {
+TEST(sysobject, read32) {
```
Comment (tswhison) [inline tests/unit/gtsys.cpp]: More tests coming in a future commit?

Comment (r-rojo) [reply tests/unit/gtsys.cpp]: Yes, more tests coming

**Comment by r-rojo** [comment]
@asgibbons, I changed the ReadBytes function to use partial read.
According to Linux kernel documentation on sysfs, doing a partial read with offset of 0 will cause the kernel driver to fill its buffer. Would it be useful to include a flag in the ReadBytes function that will force a buffer reload by seeking to zero?

File: libopae/src/sysobject.c [hunk 4]
```diff
@@ -98,18 +101,42 @@ fpga_result __FPGA_API__ fpgaReadObjectBytes(fpga_token token, const char *key,
...
 	if (res) {
 		return res;
 	}
 
+	fd = open(objpath, O_RDONLY);
+	if (fd < 0) {
+		return FPGA_NOT_FOUND;
+	}
+
+	count = pread(fd, buffer, *len, offset);
```
Comment (tswhison) [inline libopae/src/sysobject.c]: Why not loop here if the call fails and errno == EINTR?  man pread/pwrite says that the fn's can fail and set any error code that read/write can set.

File: libopae/src/sysobject.c [hunk 5]
```diff
@@ -144,17 +171,41 @@ fpga_result __FPGA_API__ fpgaWriteObject64(fpga_handle handle, const char *key,
...
+	res = cat_handle_sysfs_path(objpath, handle, key);
 	if (res) {
 		return res;
 	}
+	fd = open(objpath, O_WRONLY);
+	if (fd < 0) {
+		return FPGA_NOT_FOUND;
+	}
+
+	count = pwrite(fd, buffer, *len, offset);
```
Comment (tswhison) [inline libopae/src/sysobject.c]: Same comment as above.  Loop when pwrite fails with EINTR.

File: libopae/src/sysobject.c [hunk 6]
```diff
@@ -146,15 +173,84 @@ fpga_result __FPGA_API__ fpgaWriteObjectBytes(fpga_handle handle,
...
+
+
+	rw_buffer = calloc(fsize, sizeof(char));
+	if (rw_buffer == NULL) {
+		res = FPGA_NO_MEMORY;
+		goto out_close;
+	}
+
+	// copy bytes to write to rw_write buffer
+	memcpy(rw_buffer + offset, buffer, len);
```
Comment (tswhison) [inline libopae/src/sysobject.c]: probably memcpy_s

**Comment by asgibbons** [comment]
@r-rojo The scenario is this:  I read a header (4-10 bytes) that contains the length of the record following.  I then read the first structure (size is count from header) immediately following the header.  Then (immediately after the record) there is another header with size.  This continues until EOF.  With ReadBytes, there will be an open, seek, read, close sequence for each block, which could be very expensive.

Any way I could just get an fd returned?  I could then do reads at will.

**Comment by asgibbons** [comment]
A strong argument can be made that this *does* comply - the SDR sysfs file contains an array of sensor description records, and the sensors file contains an array of sensor readings.  Each of these files is an array of the same type.  The only sticky part is that the type has variable-length constituents.  This is not unlike the sysfs file ["/sys/devices/system/node/node0/numastat"](/sys/devices/system/node/node0/numastat) which displays an array of statistics for NUMA interactions.  The lines within this file are variable-length, but all of the same "type."

I guess I'd be OK with the API returning the entire file contents, which I could subsequently parse.  It should probably have a mechanism to return the size of the buffer required so we can be efficient with memory allocation.  It should read the minimum of the <number of bytes in the file> or <the passed-in size>.

**Comment by r-rojo** [comment]
@asgibbons, Hmm... I guess it's sort of a grey area... having multiple headers within one file is the tricky part. So as far as this API goes, what do you think should be the agreement?
* Caller allocates and frees buffer
  This requires the caller to have some method in the API to determine how much to allocate
* API allocates and frees buffer
  The API function ReadBytes allocates the buffer and returns the size in the output parameter.
  Another API function is made to free
* API allocates but caller frees
  The API function ReadBytes allocates the buffer and returns the size in the output parameter.
  The caller simply calls free on the buffer

Another idea is that we could introduce an additional opaque type to represent device attributes. 
```C
fpgaGetAttribute(token|handle, key, fpga_attribute*);
fpgaDestroyAttribute(tfpga_attribute*);
fpgaReadAttribute64(fpga_attribute, uint64_t*);
fpgaReadAttributeByes(fpga_attribute, uint8_t*, count, offset);
```

Thoughts?

**Comment by asgibbons** [comment]
@r-rojo I would prefer option 1, as I believe it's best for the caller to manage the memory.  I would recommend we follow fpgaEnumerate's lead - fpgaReadObjectBytes() gets passed a NULL buffer pointer and it returns the file size perhaps?

File: common/include/opae/sysobject.h [hunk 7]
```diff
@@ -80,15 +80,20 @@ fpga_result fpgaReadObject64(fpga_token token, const char *key,
  *
  * @param[in] token Token identifying a resource (accelerator or device)
  * @param[in] key A key identifying the object with respect to the resource
- * @param[out] buffer Pointer to memory to store the object's value
+ * @param[out] buffer Pointer to memory to store the object's bytes - a null
+ * pointer indicates a request for the total number of bytes required
  * @param[in] offset Start of byte sequence to read
- * @param[in] len The number of bytes to read. This is updated with the
+ * @param[out] len The number of bytes to read. This is updated with the
```
Comment (asgibbons) [inline common/include/opae/sysobject.h]: This is actually '[in/out]'.

File: libopae/src/sysobject.c [hunk 8]
```diff
@@ -41,6 +45,38 @@
 		}                                                              \
 	} while (false);
 
+static inline ssize_t eintr_pread(int fd, void *buf, size_t count, size_t offset)
+{
+	ssize_t bytes_read = 0, total_read = 0;
+	char *ptr = buf;
+	while (total_read < (ssize_t)count) {
+		bytes_read = pread(fd, ptr + total_read, count, offset);
```
Comment (tswhison) [inline libopae/src/sysobject.c]: Don't you need..
  count - total_read
  offset + total_read

File: libopae/src/sysobject.c [hunk 9]
```diff
@@ -41,6 +45,38 @@
...
+	}
+	return total_read;
+}
+
+static inline ssize_t eintr_write(int fd, void *buf, size_t count)
+{
+	ssize_t bytes_written = 0, total_written = 0;
+	char *ptr = buf;
+	while (total_written < (ssize_t)count) {
+		bytes_written = write(fd, ptr + total_written, count);
```
Comment (tswhison) [inline libopae/src/sysobject.c]: Don't you need..
  count - total_written

**Pull request merged**